src/features/testingProvider.ts

```diff
@@ -0,0 +1,983 @@
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from "vscode";
+import CompositeDisposable from "../CompositeDisposable";
+import AbstractProvider from "./abstractProvider";
+import OptionProvider from "../observers/OptionProvider";
+import { LanguageMiddlewareFeature } from "../omnisharp/LanguageMiddlewareFeature";
+import {
+    MSBuildProject,
+    ProjectInformationResponse,
+    TargetFramework,
+    V2,
+} from "../omnisharp/protocol";
+import { OmniSharpServer } from "../omnisharp/server";
+import * as serverUtils from "../omnisharp/utils";
+import TestManager from "./dotnetTest";
+import { Subject } from "rxjs";
+import {
+    buffer,
+    debounceTime,
+    distinct,
+    filter,
+    mergeMap,
+} from "rxjs/operators";
+import Disposable from "../Disposable";
+import { CancellationTokenSource } from "vscode-languageserver-protocol";
+import { EventStream } from "../EventStream";
+import {
+    DotNetTestDiscoveryResult,
+    DotNetTestDiscoveryStart,
+} from "../omnisharp/loggingEvents";
+import Structure = V2.Structure;
+import SymbolKinds = V2.SymbolKinds;
+import SymbolPropertyNames = V2.SymbolPropertyNames;
+import SymbolRangeNames = V2.SymbolRangeNames;
+
+export default class TestingProvider extends AbstractProvider {
+    private readonly _testAssemblies: Map<string, TestAssembly> = new Map<
+        string,
+        TestAssembly
+    >();
+
+    public controller: vscode.TestController;
+
+    private readonly _fileChangeDebouncer = new Subject<string>();
+
+    private readonly _projectChangedDebouncer = new Subject<MSBuildProject>();
+
+    constructor(
+        private readonly _optionProvider: OptionProvider,
+        private readonly _eventStream: EventStream,
+        public readonly testManager: TestManager,
+        server: OmniSharpServer,
+        languageMiddlewareFeature: LanguageMiddlewareFeature
+    ) {
+        super(server, languageMiddlewareFeature);
+        this.controller = vscode.tests.createTestController(
+            "ms-dotnettools:csharp",
+            ".Net Test Explorer"
+        );
+        const flushFiles = this._fileChangeDebouncer.pipe(debounceTime(1500));
+        const s1 = this._fileChangeDebouncer
+            .pipe(
+                filter((x) => x.endsWith(".cs")),
+                distinct(null, flushFiles),
+                buffer(flushFiles),
+                filter((x) => x.length > 0),
+                mergeMap((files) => this._reportFileChanges(files))
+            )
+            .subscribe();
+        const flushProjects = this._projectChangedDebouncer.pipe(
+            debounceTime(1500)
+        );
+        const s2 = this._projectChangedDebouncer
+            .pipe(
+                // TODO this is obviously a bad idea
```

Comment:
There could be a option for this. Like a glob pattern or something similar. 
This avoid loading every project on startup

Reply:
We could also consider a new O# request to get the list of Test projects.

Reply:
or we could add the project path to the test discovery response https://github.com/OmniSharp/omnisharp-roslyn/blob/master/src/OmniSharp.DotNetTest/Models/DiscoverTestsResponse.cs

Reply:
@filipw Is there already any code that marks projects as test projects (e.g. based on package references, <IsTestProject> or name)?